[WIP] Upgradability: contracts+tests

vyomshm opened the pull request:
Implemented Inherited Storage Proxy pattern for upgrading token+crowdsale contracts

ref #102

vyomshm commented:
Also, I had a question regarding updating contract references within the DApp. According to the updated contracts, all references to the each of the `IkuToken` , `ResearchSpecificToken` and `RSTCrowdsale` contracts need to be updated to reflect their respective proxy addresses.
I believe only the following files contain such references & will need to be updated -

- jobs/deploy-crowdsale.js
- jobs/finalize-crowdsale.js
- src/utils/IkuTokenHelper.js

Am I on the right track or are there any other files which reference the contracts?

ikuteer commented:
@vyomshm  regarding the database, you need to declare `$DATABASE_URL` pointing to your local PGSQL instance. then you can run `yarn db:setup`

Regarding the contracts references,  there are a few more places:
- [ResearchSpecificToken](https://github.com/ikunetwork/alpha/search?q=ResearchSpecificToken&unscoped_q=ResearchSpecificToken)
- [IkuToken](https://github.com/ikunetwork/alpha/search?q=IkuToken&unscoped_q=IkuToken)

Basically what you mentioned + migrations + api/ references + tests

vyomshm commented:
@ikuteer , I've added all the needed contracts and written all the relevant tests. 
Now both of your token contracts i.e. `IkuToken` & `ResearchSpecificToken` are upgradable. This means that you can deploy new versions of your contracts and they'll work with the same proxy addresses created with the earlier versions.

Contract deployment and upgradation process is almost similar to that mentioned in the blog you referenced. I've written extensive tests to demonstrate how contracts should be deployed, initialized and upgraded(see `IkuToken_v2` and `ResearchSpecificToken_v2` and added test files).

All upgraded contracts -
- should inherit from the previous versions
- should include an `initializer` function for setting constructor variables(if any) + `isInitialized` flag variable to keep track

I've also included tests for your `RSTCrowdsale` contract using `ResearchSpecificToken`'s proxy addresses.

Proxy addresses will need to be saved when deploying initial versions of the contracts & these can then be referenced everywhere. Whenever you deploy an upgraded version, call `upgradeTo` function on the `registry` contract and the initialize it by calling the `initializer`, if any.

Status-

- [x] test passing - 100 %

Let me know if you have any doubts. 
ref #102

vyomshm commented:
@ikuteer have you had a chance to review? Should I "submit work" on gitcoin page?

ikuteer commented on api/Database/index.js:
```diff
@@ -3,10 +3,12 @@ const { Client } = require('pg');
+      //   connectionString: process.env.DATABASE_URL,
+      //   ssl: true,
+      // });
+      const client = new Client();
```
Please revert this change

vyomshm replied:
I'll fix this.

ikuteer commented on contracts/RSTCrowdsale.sol:
```diff
@@ -29,4 +28,4 @@ contract RSTCrowdsale is CappedCrowdsale, RefundableCrowdsale, MintedCrowdsale {
   }
 
-}
+}
```
Restore the changes on this file

ikuteer replied:
Also, this contract should be upgradeable too but looks like you missed it.

vyomshm replied:
@ikuteer , I'm sorry I should have clarified. This crowdsale contract is not upgradable using any proxy pattern. One of the contract that `RSTCrowdsale` inherits is `RefundableCrowdsale`. This contract has a private variable `escrow` which cannot be set using the proxy pattern architecture. I even consulted the zeppelin os developers to ask about this. Here's a screeshot of my interaction - 
![screen shot 2018-09-02 at 3 26 18 pm](https://user-images.githubusercontent.com/20184946/44954739-9f0c3300-aec4-11e8-9f10-5b7c1139828e.png)

Since it might not be safe to tamper the internal workings of the crowdsale architecture & 
 `RSTCrowdsale` contract will only be ever used once during the crowdsale, I would recommend not trying to make it upgradable.

The current setup is like this- you deploy a `ResearchSpecificToken` using proxy pattern, this will give you a proxy address to this token, you can then use this proxy token address inside the original `RSTCrowdsale` contract.

ikuteer commented on migrations/2_deploy_contracts.js:
```diff
@@ -1,55 +1,63 @@
+	}).then((instance)=>{
+		researchSpecificToken = instance;
+		console.log('initializing ResearchSpecificToken ...');
+		return researchSpecificToken.initializeRSToken(18, 'ResearchSpecificToken', 'RST')
```
The ResearcSpecificToken should be deployed only via jobs. So this shouldn't be here.

ikuteer commented on migrations/3_upgrade_contracts.js:
```diff
@@ -0,0 +1,81 @@
+
+	//deploy all version-2 contracts
+	deployer.deploy(IkuToken_v2);
+	deployer.deploy(
```
Same as my previous comment

ikuteer commented on src/components/Tabs.js:
```diff
@@ -8,7 +8,9 @@ const CustomTabs = props => (
-    {this.props.children.map((el, i) => <TabPanel>{el.props.title}</TabPanel>)}
+    {this.props.children.map((el, i) => (
+      <TabPanel>{el.props.title}</TabPanel>
+    ))}
```
Looks like you ran prettier or something like that in the entire code base. This shouldn't be part of this PR.

vyomshm replied:
Yeah, I ran the linter script as specified in the `package.json` (i.e. `yarn lint::all::fix`) before pushing the commits. It automatically made a lot of changes in multiple files, files I hadn't even touched. But I'm sure none of the codebase was affected.

ikuteer commented on src/pages/proposal/Proposal.js:
```diff
@@ -506,7 +506,7 @@ class Proposal extends Component {
   renderFundedPercentage() {
     const percentage = Math.ceil(
-      this.state.amount_raised * 100 / this.state.funds_required
+      (this.state.amount_raised * 100) / this.state.funds_required
```
Same on this file

ikuteer commented on test/IkuToken.js:
```diff
@@ -1,6 +1,7 @@
 import decodeLogs from './helpers/decodeLogs';
 
 const IkuToken = artifacts.require('IkuToken');
+const IkuToken_v2 = artifacts.require('IkuToken_v2');
```
What's the point of requiring IkuToken_v2 here if you're not doing anything with it?

vyomshm replied:
nice catch! I've removed it.

vyomshm commented:
#118 please refer to my new PR. I'm closing this one
